les/client_handler.go

```diff
@@ -20,416 +20,509 @@ import (
+		p.chainRecent = 0
+	}
+	if recv.get("serveStateSince", &p.stateSince) != nil {
+		return errResp(ErrUselessPeer, "peer cannot serve requests")
+	}
+	if recv.get("serveRecentState", &p.stateRecent) != nil {
+		p.stateRecent = 0
+	}
+	if recv.get("txRelay", nil) != nil {
+		return errResp(ErrUselessPeer, "peer cannot serve requests")
+	}
+	if p.version >= lpv4 {
+		var recentTx uint
+		if err := recv.get("recentTxLookup", &recentTx); err != nil {
+			return err
+		}
+		p.txHistory = uint64(recentTx)
+	} else {
+		// The weak assumption is held here that legacy les server(les2,3)
+		// has unlimited transaction history. The les serving in these legacy
+		// versions is disabled if the transaction is unindexed.
+		p.txHistory = txIndexUnlimited
 	}
-	handler.fetcher = newLightFetcher(backend.blockchain, backend.engine, backend.peers, handler.ulc, backend.chainDb, backend.reqDist, handler.synchronise)
-	handler.downloader = downloader.New(height, backend.chainDb, backend.eventMux, nil, backend.blockchain, handler.removePeer)
-	handler.backend.peers.subscribe((*downloaderPeerNotify)(handler))
-	return handler
-}
 
-func (h *clientHandler) start() {
-	h.fetcher.start()
+	recv.get("checkpoint/value", &p.checkpoint)
+	recv.get("checkpoint/registerHeight", &p.checkpointNumber)
+	return nil
 }
 
-func (h *clientHandler) stop() {
-	close(h.closeCh)
-	h.downloader.Terminate()
-	h.fetcher.stop()
-	h.wg.Wait()
-}
+// peerConnected implements connectionModule
+func (h *clientHandler) peerConnected(p *peer) (func(), error) {
+	/*if h.peers.len() >= h.blockchain.Config().LightPeers && !p.Peer.Info().Network.Trusted { //TODO ???
+		return nil, p2p.DiscTooManyPeers
+	}*/
+	serverConnectionGauge.Update(int64(h.peers.len()))
+	// Discard all the announces after the transition
+	// Also discarding initial signal to prevent syncing during testing.
+	if h.fetcher != nil && !h.noInitAnnounce {
+		h.fetcher.announce(p, &announceData{Hash: p.headInfo.Hash, Number: p.headInfo.Number, Td: p.headInfo.Td})
+	}
 
-// runPeer is the p2p protocol run function for the given version.
-func (h *clientHandler) runPeer(version uint, p *p2p.Peer, rw p2p.MsgReadWriter) error {
-	trusted := false
-	if h.ulc != nil {
-		trusted = h.ulc.trusted(p.ID())
-	}
-	peer := newServerPeer(int(version), h.backend.config.NetworkId, trusted, p, newMeteredMsgWriter(rw, int(version)))
-	defer peer.close()
-	h.wg.Add(1)
-	defer h.wg.Done()
-	err := h.handle(peer, false)
-	return err
+	return func() {
+		p.fcServer.DumpLogs()
+		serverConnectionGauge.Update(int64(h.peers.len()))
+	}, nil
 }
 
-func (h *clientHandler) handle(p *serverPeer, noInitAnnounce bool) error {
-	if h.backend.peers.len() >= h.backend.config.LightPeers && !p.Peer.Info().Network.Trusted {
-		return p2p.DiscTooManyPeers
+// messageHandlers implements messageHandlerModule
+func (h *clientHandler) messageHandlers() messageHandlers {
+	return messageHandlers{
+		messageHandlerWithCodeAndVersion{
```

Comment:
IIUC, with the version indicator, we can define different handlers for the same message code but different versions? 

For example, both `handleAnnounce` for `LES4` and `handleAnnounceV5` for `LES5` can be defined in the `messageHandlers`. It will make upgrading protocol a lot of easier I think.